backends/rollup/rollup-init-transaction/Main.hs

```diff
@@ -0,0 +1,116 @@
+import           Cardano.Api.Ledger                    (toCBOR)
+import           Cardano.Api.Shelley                   (File (..), PlutusScript (..), fromPlutusData,
+                                                        scriptDataToJsonDetailedSchema)
+import           Codec.CBOR.Write                      (toStrictByteString)
+import           Control.Monad                         (void)
+import           Data.Aeson                            (encode)
+import qualified Data.Aeson                            as Aeson
+import qualified Data.ByteString                       as BS
+import qualified Data.ByteString.Lazy                  as BL
+import qualified PlutusLedgerApi.V2                    as V2
+import qualified PlutusLedgerApi.V3                    as V3
+import           PlutusTx                              (CompiledCode, ToData (..))
+import           Prelude                               (Bool (..), Either (..), FilePath, IO, Integer, Maybe (..),
+                                                        Show (..), head, print, putStr, read, return, ($), (++), (.),
+                                                        (<$>))
+import           System.Directory                      (createDirectoryIfMissing)
+import           System.Environment                    (getArgs)
+import           Test.QuickCheck.Arbitrary             (Arbitrary (..))
+import           Test.QuickCheck.Gen                   (generate)
+import           Text.Parsec                           (many1, parse)
+import           Text.Parsec.Char                      (digit)
+import           Text.Parsec.String                    (Parser)
+
+import           ZkFold.Cardano.Examples.EqualityCheck (equalityCheckVerificationBytes)
+import           ZkFold.Cardano.OffChain.E2E           (EqualityCheckContract (..))
+import           ZkFold.Cardano.UPLC                   (parkingSpotCompiled, rollupCompiled')
+import           ZkFold.Cardano.UPLC.Rollup            (RollupRedeemer (..))
+
+
+saveRollupPlutus :: IO ()
+saveRollupPlutus = do
+  x           <- generate arbitrary
+  ps          <- generate arbitrary
+  targetValue <- generate arbitrary
+
+  let contract = EqualityCheckContract x ps targetValue
+
+  BL.writeFile "../../test-data/plonk-raw-contract-data.json" $ encode contract
+
+  putStr $ "x: " ++ show x ++ "\n" ++ "ps: " ++ show ps ++ "\n" ++ "targetValue: " ++ show targetValue ++ "\n\n"
+
+  let (ledgerRules, iniState, proof) = equalityCheckVerificationBytes x ps targetValue
+
+  let redeemerRollupA = RollupRedeemer
+        { rrProof   = proof
+        , rrAddress = V3.Address rollupCredential Nothing
+        , rrValue   = lovelace 3000000
+        , rrState   = iniState
+        , rrUpdate  = [iniState]
+        }
+        where
+          rollupCredential = credentialOf $ rollupCompiled' ledgerRules
+          lovelace         = V2.singleton V2.adaSymbol V2.adaToken
+
+  savePlutus "../../assets/rollup.plutus" $ rollupCompiled' ledgerRules
```

Comment:
I don't understand why are we using `rollupCompiled'` for the actual rollup transactions. I thought we were only using it for fee calculation, but it does not seem to be the case.

Reply:
Well, first rollup verification would indeed work with `rollupCompiled`, since `iniState` in
```haskell
  let (ledgerRules, iniState, proof) = emptyCircuitVerificationBytes x ps
```
is passed as the initial state.  Since state is subsequently updated, the proof `proof` will no longer verify.

Reply:
Can we honestly compute proofs for every rollup batch transactions? That's what we will be doing anyway.

Reply:
As I see it, the issue is that computing the proof with 
```haskell
  let (ledgerRules, iniState, proof) = emptyCircuitVerificationBytes x ps
```
makes `proof` (and the validator's parameter `ledgerRules`) depend on the random choices of `x` and `ps`, not on "state".  So, to honestly compute proofs we would need the actual circuit, not "emptyCircuit", I think.  On the other hand, using `rollupCompiled'` at least simulates the computational effort of verifying the proof, which might be enough for a prototype.